pkg/client/client_v2_test.go

```diff
@@ -0,0 +1,696 @@
+								ConfigStateIdx: 1,
+								Config:         "config_unit_one",
+							},
+							{
+								Id:             unitTwo.id,
+								Type:           proto.UnitType_INPUT,
+								State:          proto.State_HEALTHY,
+								ConfigStateIdx: 1,
+								Config:         "config_unit_two",
+							},
+						},
+					}
+				} else if (unitOne.state == UnitStateHealthy && unitTwo.state == UnitStateHealthy) || (unitOne.state == UnitStateHealthy && unitTwo.state == UnitStateStopping) {
+					// stop second input
+					return &proto.CheckinExpected{
+						Units: []*proto.UnitExpected{
+							{
+								Id:             unitOne.id,
+								Type:           proto.UnitType_OUTPUT,
+								State:          proto.State_HEALTHY,
+								ConfigStateIdx: 1,
+								Config:         "",
+							},
+							{
+								Id:             unitTwo.id,
+								Type:           proto.UnitType_INPUT,
+								State:          proto.State_STOPPED,
+								ConfigStateIdx: 1,
+								Config:         "",
+							},
+						},
+					}
+				} else if unitOne.state == UnitStateHealthy && unitTwo.state == UnitStateStopped {
+					// input stopped, remove the unit
+					return &proto.CheckinExpected{
+						Units: []*proto.UnitExpected{
+							{
+								Id:             unitOne.id,
+								Type:           proto.UnitType_OUTPUT,
+								State:          proto.State_HEALTHY,
+								ConfigStateIdx: 1,
+								Config:         "",
+							},
+						},
+					}
+				} else {
+					// outside of the state diagram we are testing
+					panic(fmt.Errorf("invalid state; unitOne(%v) - unitTwo(%v)", unitOne.state, unitTwo.state))
+				}
+			}
+			// disconnect
+			return nil
+		},
+		ActionImpl: func(response *proto.ActionResponse) error {
+			// actions not tested here
+			return nil
+		},
+		actions: make(chan *performAction, 100),
+	}
+	require.NoError(t, srv.Start())
+	defer srv.Stop()
+
+	var errsLock sync.Mutex
+	var errs []error
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	client := NewV2(fmt.Sprintf(":%d", srv.Port), token, VersionInfo{}, grpc.WithTransportCredentials(insecure.NewCredentials())).(*clientV2)
+	storeErrors(ctx, client, &errs, &errsLock)
+
+	// receive the units
+	var unitsLock sync.Mutex
+	units := make(map[string]*Unit)
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case change := <-client.UnitChanges():
+				switch change.Type {
+				case UnitChangedAdded:
+					unitsLock.Lock()
+					units[change.Unit.ID()] = change.Unit
+					unitsLock.Unlock()
+					change.Unit.UpdateState(UnitStateHealthy, "Healthy", map[string]interface{}{
+						"custom": "payload",
+					})
+				case UnitChangedModified:
+					state, _ := change.Unit.Expected()
+					if state == UnitStateStopped {
+						change.Unit.UpdateState(UnitStateStopping, "Stopping", nil)
+						go func() {
+							time.Sleep(100 * time.Millisecond)
+							change.Unit.UpdateState(UnitStateStopped, "Stopped", nil)
+						}()
```

Comment:
Should we make the "Healthy", "Stopping", "Stopped" and other a public constant?

Reply:
For testing make it constant? For the library, no I would prefer they provide more detail than that. Like "Connected to elasticsearch(https://localhost:8200)" or something.

Reply:
OK good point.